// jira-query/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _init2 = require('./init');

var _init3 = _interopRequireDefault(_init2);

var _config = require('./config');

var _jira = require('./jira');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var api = {
  config: _config.setConfig
};

var initPromise = void 0;

function init() {
  initPromise = initPromise || (0, _init3.default)();
  return initPromise;
}

Object.assign(api, {
  getMyIssues: function getMyIssues() {
    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return init().then(function () {
      return _jira.getMyIssues.apply(undefined, args);
    });
  },
  getMyOpenIssues: function getMyOpenIssues() {
    for (var _len2 = arguments.length, args = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
      args[_key2] = arguments[_key2];
    }

    return init().then(function () {
      return _jira.getMyOpenIssues.apply(undefined, args);
    });
  },
  getIssue: function getIssue() {
    for (var _len3 = arguments.length, args = Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
      args[_key3] = arguments[_key3];
    }

    return init().then(function () {
      return _jira.getIssue.apply(undefined, args);
    });
  },
  jql: function jql() {
    for (var _len4 = arguments.length, args = Array(_len4), _key4 = 0; _key4 < _len4; _key4++) {
      args[_key4] = arguments[_key4];
    }

    return init().then(function () {
      return _jira.searchForIssues.apply(undefined, args);
    });
  }
});

exports.default = api;
